govcd/cse.go

```diff
@@ -0,0 +1,379 @@
+func (cluster *CseKubernetesCluster) UpgradeCluster(kubernetesTemplateOvaId string, refresh bool) error {
+	return cluster.Update(CseClusterUpdateInput{
+		KubernetesTemplateOvaId: &kubernetesTemplateOvaId,
+	}, refresh)
+}
+
+// SetNodeHealthCheck executes an update on the receiver cluster to enable or disable the machine health check capabilities.
+// If refresh=true, it retrieves the latest state of the cluster from VCD before updating.
+func (cluster *CseKubernetesCluster) SetNodeHealthCheck(healthCheckEnabled bool, refresh bool) error {
+	return cluster.Update(CseClusterUpdateInput{
+		NodeHealthCheck: &healthCheckEnabled,
+	}, refresh)
+}
+
+// SetAutoRepairOnErrors executes an update on the receiver cluster to change the flag that controls the auto-repair
+// capabilities of CSE. If refresh=true, it retrieves the latest state of the cluster from VCD before updating.
+// NOTE: This method can only be used in CSE versions < 4.1.1
+func (cluster *CseKubernetesCluster) SetAutoRepairOnErrors(autoRepairOnErrors bool, refresh bool) error {
+	return cluster.Update(CseClusterUpdateInput{
+		AutoRepairOnErrors: &autoRepairOnErrors,
+	}, refresh)
+}
+
+// Update executes an update on the receiver CSE Kubernetes Cluster on any of the allowed parameters defined in the input type.
+// If refresh=true, it retrieves the latest state of the cluster from VCD before updating.
+func (cluster *CseKubernetesCluster) Update(input CseClusterUpdateInput, refresh bool) error {
+	if refresh {
+		err := cluster.Refresh()
+		if err != nil {
+			return err
+		}
+	}
+
+	if cluster.State == "" {
+		return fmt.Errorf("can't update a Kubernetes cluster that does not have any state")
+	}
+	if cluster.State != "provisioned" {
+		return fmt.Errorf("can't update a Kubernetes cluster that is not in 'provisioned' state, as it is in '%s'", cluster.capvcdType.Status.VcdKe.State)
+	}
+
+	if input.AutoRepairOnErrors != nil {
+		// Since CSE 4.1.1, the AutoRepairOnError toggle can't be modified and is turned off
+		// automatically by the CSE Server.
+		v411, err := semver.NewVersion("4.1.1")
+		if err != nil {
+			return fmt.Errorf("can't update the  'Auto Repair on Errors' flag: %s", err)
+		}
+		if cluster.CseVersion.GreaterThanOrEqual(v411) {
+			return fmt.Errorf("the 'Auto Repair on Errors' flag can't be changed after the cluster is created since CSE 4.1.1")
+		}
+		cluster.capvcdType.Spec.VcdKe.AutoRepairOnErrors = *input.AutoRepairOnErrors
+	}
+
+	updatedCapiYaml, err := cluster.updateCapiYaml(input)
+	if err != nil {
+		return err
+	}
+	cluster.capvcdType.Spec.CapiYaml = updatedCapiYaml
+
+	marshaledPayload, err := json.Marshal(cluster.capvcdType)
+	if err != nil {
+		return err
+	}
+	entityContent := map[string]interface{}{}
+	err = json.Unmarshal(marshaledPayload, &entityContent)
+	if err != nil {
+		return err
+	}
+
+	// We do this loop to increase the chances that the Kubernetes cluster is successfully updated, as this method will be
+	// "fighting" with the CSE Server
```

Comment:
To expand on the comment - so the point is that CSE server itself will perform updates and `etag` keeps on changing, while we must also squeeze in our update, right?

Reply:
Improved the comment a bit.